Nwfix

nefelin opened the pull request:
Made slight modifications adding CompositionDetails so that xaqt can feed boilerplate and comment char info to clients.

Is currently necessary for nodewars to function properly

nefelin committed:
Added composition details so /languages/ route now hands out a
map[languagename(string)]CompositionDetails

nefelin committed:
fixed paths so assuming GOPATH is set and project not a fork, xaqt is callable from anywhere

frenata commented on cmd/xaqt/main.go:
```diff
@@ -32,7 +33,7 @@ var context *xaqt.Context
 	port := getEnv("XAQT_PORT", "31337")
 
-	compilers := xaqt.ReadCompilers("data/compilers.json")
+	compilers := xaqt.ReadCompilers(filepath.Join(xaqt.DataPath(), "compilers.json"))
```
This is a solid default compared to the prior behavior. In the longer run I'd want there to be an optional CLI flag to manually specify this location and/or a REST endpoint to re-specify the compilers details while the server is running.

frenata commented on compilers.go:
```diff
@@ -18,7 +19,19 @@ type ExecutionDetails struct {
 
-// Reads a compilers map from a file.
+// CompositionDetails specifies how to write code in a given language
+type CompositionDetails struct {
```
I think this is a good split, separates the concerns more than in the original cut, but preserves the knowledge you need.

frenata commented on context.go:
```diff
@@ -70,7 +70,7 @@ func (c *Context) run(language, code, stdinGlob string) (string, Message) {
 	}
 
-	sb := newSandbox(lang, code, stdinGlob, c.options)
+	sb := newSandbox(lang.ExecutionDetails, code, stdinGlob, c.options)
```
This might have even worked without this change, via embedded things being accessible seamlessly, but it never hurts to be explicit. :)

frenata commented on compilers.go:
```diff
@@ -36,13 +49,14 @@ func ReadCompilers(filename string) Compilers {
 	for k, v := range c {
 		if v.Disabled != "true" {
-			langs = append(langs, k)
+			langs[k] = v.CompositionDetails
```
👍

frenata commented on options.go:
```diff
@@ -14,10 +14,19 @@ import (
+func DataPath() string {
+	gopath, ok := os.LookupEnv("GOPATH")
+	if !ok {
+		panic("Error locating data path, go GOPATH set")
```
Did you try this? I think `panic` will pop a full stack trace. `log.Fatalf` may be preferable for an end-user oriented message.

frenata replied:
Fixed myself along with some other stuff related to logging once I got in there.

frenata commented on context.go:
```diff
@@ -86,4 +86,4 @@ func (c *Context) run(language, code, stdinGlob string) (string, Message) {
 
 // Languages returns a list of available language names.
-func (c *Context) Languages() []string { return c.compilers.availableLanguages() }
+func (c *Context) Languages() map[string]CompositionDetails { return c.compilers.availableLanguages() }
```
It might be nice to expose endpoints that return both the raw list of languages and the composition details.

frenata reviewed (changes requested):
The only thing I think needs to be fixed is the `panic`.

frenata committed:
Avoid panic when 'GOPATH' not set.

A fatal log message is printed and the program is exited instead.
Tweaked existing env-var log messages to be clear that they are only
informational.

frenata committed:
Don't crash the server on a bad /evaluate/ request

Instead, return at least a simple error response, internally log more
detail, and wait for the next request.

I picked up on this problem while checking out existing `panic`s in
order to align other user-facing log messages to the previous changes.
The user returned message could be better, but this much superior the
the previous behavior of just blowing up entirely.

There are no more plain `panic`s now, and all remaining unadorned
log.Fatal's are IO related.
